errors: improve error creation performance

BridgeAR opened the pull request:
This is a first step to significantly faster NodeErorrs. The code itself should also be cleaned up in a following PR but it already became bigger as I worked upon different parts related to errors in Node.js.

This does change stack traces for users in few cases: originally we added extra stack frames to compensate for stack frames we want to hide. The mechanism as it was implemented had a high cost on all Node.js errors in case they were created in a deeply nested function. Stack frames that are removed are now not going to be compensated for anymore. I am going to think about other ways to do this later on again.

Refs: https://github.com/nodejs/performance/issues/40

```
                                                                        confidence improvement accuracy (*)   (**)   (***)
error/hidestackframes.js nested=0 n=10000 type='direct-call-noerr'                      0.52 %       ±5.47% ±7.30%  ±9.56%
error/hidestackframes.js nested=0 n=10000 type='direct-call-throw'             ***     81.31 %       ±7.21% ±9.66% ±12.72%
error/hidestackframes.js nested=0 n=10000 type='hide-stackframes-noerr'                 2.66 %       ±6.63% ±8.83% ±11.49%
error/hidestackframes.js nested=0 n=10000 type='hide-stackframes-throw'        ***     79.61 %       ±6.63% ±8.87% ±11.66%
error/hidestackframes.js nested=1 n=10000 type='direct-call-noerr'                      0.21 %       ±3.57% ±4.76%  ±6.19%
error/hidestackframes.js nested=1 n=10000 type='direct-call-throw'             ***     90.10 %       ±5.93% ±7.93% ±10.40%
error/hidestackframes.js nested=1 n=10000 type='hide-stackframes-noerr'                -2.34 %       ±4.07% ±5.42%  ±7.06%
error/hidestackframes.js nested=1 n=10000 type='hide-stackframes-throw'        ***     85.71 %       ±6.18% ±8.23% ±10.73%
error/node-error.js type='node' n=100000                                       ***    166.64 %       ±7.01% ±9.39% ±12.35%
error/node-error.js type='regular' n=100000                                            -2.29 %       ±2.98% ±3.97%  ±5.16%
```

targos commented on lib/assert.js:
```diff
@@ -278,16 +278,17 @@ function parseCode(code, offset) {
 function getErrMessage(message, fn) {
-  const tmpLimit = Error.stackTraceLimit;
-  const errorStackTraceLimitIsWritable = isErrorStackTraceLimitWritable();
+  // eslint-disable-next-line node-core/prefer-primordials
+  const { stackTraceLimit } = Error;
```
What about adding a `getStackTraceLimit` to `internal/errors` so we don't have to disable the ESLint rule everywhere?

BridgeAR replied:
Thinking about it: I am not certain that's ideal as we have to just export `Error.stackTraceLimit` from `internal/errors` in that case by wrapping it in a method. We could just improve the rule itself to skip this property.

targos replied:
Are you afraid of the perf cost of calling a method?
Having a "get" counterpart of the "set" that already exists seemed natural to me.

targos replied:
I'm also fine with an exception to the rule.

aduh95 replied:
It's already an exception on the rule: https://github.com/nodejs/node/blob/fe514bf960ca1243b71657af662e7df29f5b57cf/lib/.eslintrc.yaml#L205-L208

If it's complaining, it means there's a bug with the rule that we should fix (hopefully before landing this PR). I'll take a look.

aduh95 replied:
Can you rebase now that the ESLint rule has been fixed?

BridgeAR replied:
Done

anonrig commented:
cc @nodejs/performance

BridgeAR commented:
I checked how much potential is left for further improvements and it turns out this is almost an ideal state.

Our main overhead is due to creating subclasses. ~50% over the most simple errors without subclass. The other ~15-20% overhead are related to our error message creation. It would actually be possible to roughly cut the latter costs into half by creating a formatPreparation method that pre calculates states and caches these for the later usage. It would however be a non-trivial code part and I believe this is already pretty nice.

ronag commented:
@BridgeAR any chance to further explicitly optimise `AbortError` as that often occurs also in non exceptional logic (i.e. fast paths).

BridgeAR commented:
@ronag the current implementation for AbortErrors is pretty much the bare minimum of a regular error.

The question is what are these used for? If their stack trace is important, we can't improve them further. If not, then there's room for improvement by hiding the frames (which are ~75% of all CPU cycles for a regular error).
To me it seems like the main stack frames that would be relevant for users are mostly coming from the signal.reason. These stack frames would not be touched, only the newly created AbortError ones. I had a short discussion if they are important or not with @benjamingr and I have the feeling the answer is: it depends.
If you therefore have some concrete spots that need some love, please let me know and I am going to check if it's possible to further minimize the overhead.

jasnell commented:
I'm not going to have time to review the entire PR this week but the changes in `internal/errors.js` LGTM.

ljharb commented on lib/internal/console/constructor.js:
```diff
@@ -421,7 +421,10 @@ const consoleMethods = {
       name: 'Trace',
-      message: this[kFormatForStderr](args)
+      message: this[kFormatForStderr](args),
+      toString() {
+        return `${this.name}: ${this.message}`;
+      }
```
would things be faster if this method were shared instead of created every time trace is called? (perhaps memoized on first call)

BridgeAR replied:
I am not certain I am able to follow. I would expect that the trace call is not used in performance sensitive code and that it'll mainly log unique values (the error message could be different on each call, even if it's the same line of code that triggers it). As such, memoizing this would probably not bring a benefit and only cause memory to fill up?

ljharb replied:
i don't mean memoizing the string output - i just mean this one function, whose code is the same every time, can be reused rather than created fresh every time

BridgeAR replied:
Ah, yes, it would be a tad faster while likely negligible. I consider the readability more important in this case.

ljharb replied:
I would expect that too, but this entire PR assumes that performance of the error path is worth improving, no?

ljharb commented on lib/internal/errors.js:
```diff
@@ -120,8 +112,13 @@ const prepareStackTrace = (globalThis, error, trace) => {
+  // Do not use primordials here: we intercept user code here.
+  if (typeof error.toString === 'function' &&
+      error.toString !== MainContextObjectToString) {
+    errorString = error.toString();
```
```suggestion
    errorString = String(error);
```
otherwise it won't be guaranteed to be a string, i think?

aduh95 replied:
It would be wrong to use a check on `toString` then, as the string value could come from `[Symbol.toPrimitive]`

BridgeAR replied:
I feel like there is no 100% right or wrong for this case. I opted for falling back to the native toString call in case the return type is not a string.

ljharb commented on lib/internal/errors.js:
```diff
@@ -828,16 +765,21 @@ function setArrowMessage(err, arrowMessage) {
-        (frm) => !StringPrototypeStartsWith(frm.getFileName() || '',
-                                            'node:internal')
-      );
+      for (const frame of stackFrames) {
```
this should probably avoid for-of, both for robustness and for performance. can it use forEach?

BridgeAR replied:
The loop type won't make a big difference here. I personally like for...of loops. They are easy to read.

ljharb replied:
for-of invokes the iterator protocol, which is often slower than array iteration methods (which are identical to a classic for loop in modern v8)

aduh95 replied:
Last time I checked, `for(;;)` is still the fastest way to iterate over an array or array-like object.

ljharb replied:
@aduh95 modern v8 turns array iteration methods into the same underlying code as a for loop, altho that may not apply with primordials. either way, for-of loops should be avoided.

ljharb replied:
for-of is worse than a for loop, because it invokes the iterator protocol, which is slower and less robust.

BridgeAR replied:
I would like to keep this as is for now.

ljharb commented on lib/internal/errors.js:
```diff
@@ -828,16 +765,21 @@ function setArrowMessage(err, arrowMessage) {
+        }
+      }
+    } else {
+      for (const frame of stackFrames) {
```
same here

ljharb commented on lib/internal/errors.js:
```diff
@@ -828,16 +765,21 @@ function setArrowMessage(err, arrowMessage) {
     }
-    ArrayPrototypeUnshift(frames, error);
-    return ArrayPrototypeJoin(frames, '\n    at ');
+    result = error + result;
+    return result;
```
```suggestion
    return error + result;
```
assuming this is either the same or faster?

BridgeAR replied:
It's identical.

ljharb replied:
consider this a nit then to minimize reassignment :-)

BridgeAR replied:
Done

aduh95 commented on lib/internal/errors.js:
```diff
@@ -120,8 +112,13 @@ const prepareStackTrace = (globalThis, error, trace) => {
-  if (kIsNodeError in error) {
-    errorString = `${error.name} [${error.code}]: ${error.message}`;
+  // Do not use primordials here: we intercept user code here.
+  if (typeof error.toString === 'function' &&
+      error.toString !== MainContextObjectToString) {
+    errorString = error.toString();
+    if (errorString === '[Object object]') {
+      errorString = ErrorPrototypeToString(error);
+    }
```
We should take into account the `toString` may be a getter that returns a different value (or throw) on subsequent calls.
```suggestion
  const errorToString = error.toString;
  if (typeof errorToString === 'function' &&
      errorToString !== MainContextObjectToString) {
    errorString = FunctionPrototypeCall(errorToString);
    if (errorString === '[Object object]') {
      errorString = ErrorPrototypeToString(error);
    }
```

BridgeAR replied:
I already thought about it and I tested it as well: if a new error is thrown during the prepare step, that error is going to take over and is reported instead. That should exactly be the behavior that we want in this case. As such, I went for the simple solution.

aduh95 replied:
What about the following case:

```js
const error = new Error;
let hasAlreadyBeenCalled = false;
Reflect.defineProperty(error, 'toString', {
  __proto__: null,
  get() {
    if (!hasAlreadyBeenCalled) {
      hasAlreadyBeenCalled = true;
      return () => {};
    }
  },
});
throw error;
```

The first check checks one value, the other checks check another value, resulting in broken code. By desctructuring the value, you remove that footgun.

BridgeAR replied:
I added your suggestion.

ljharb commented on lib/internal/source_map/prepare_stack_trace.js:
```diff
@@ -22,30 +22,36 @@ const {
+  // Do not use primordials here: we intercept user code here.
+  if (typeof error.toString === 'function' &&
+      error.toString !== MainContextObjectToString) {
+    errorString = error.toString();
```
```suggestion
    errorString = `${error}`;
```

ljharb replied:
i think this change still needs to be made?

BridgeAR replied:
This does not apply anymore.

aduh95 commented on lib/internal/errors.js:
```diff
@@ -120,8 +112,13 @@ const prepareStackTrace = (globalThis, error, trace) => {
+  if (typeof error.toString === 'function' &&
+      error.toString !== MainContextObjectToString) {
+    errorString = error.toString();
+    if (errorString === '[Object object]') {
```
Wouldn't it be better to have another `errorToString !== MainContextObjectPrototypeToString`?

BridgeAR replied:
You mean to have a variable with the value assigned instead of using a "magic string"? I have no strong opinion either way.

targos commented on lib/internal/streams/readable.js:
```diff
@@ -370,10 +370,10 @@ Readable.prototype.setEncoding = function(enc) {
 
 // Don't raise the hwm > 1GB.
-const MAX_HWM = 0x40000000;
+const MAX_HWM = 2 ** 30;
```
I think V8 can constant-fold this expression so the perf should be identical

BridgeAR replied:
Yes, that was just a change for me for readability reasons. I can't compute hex numbers. Two to the power of n is much easier for me :D.

BridgeAR commented on lib/internal/test_runner/utils.js:
```diff
@@ -65,7 +64,7 @@ function createDeferredCallback() {
 
 function isTestFailureError(err) {
-  return err?.code === 'ERR_TEST_FAILURE' && kIsNodeError in err;
+  return err instanceof ERR_TEST_FAILURE;
```
Likely

ljharb replied:
then could it be modified to not be?

BridgeAR replied:
The alternative is actually easier to confuse: just checking for a string as name is very error prone. Changing the instance would however require quite some work to do and seems less error prone. I therefore opt for keeping this check as it is.

ljharb replied:
If node is the only creator of ERR_TEST_FAILURE errors, then it could keep a WeakSet of all the instances created, and this check could test for membership in that WeakSet. It seems like a 2-4 line change to make it zero error prone.

mcollina reviewed (approved):
lgtm, this is truly amazing

RafaelGSS reviewed (approved):
RSLGTM. Awesome work!

VoltrexKeyva reviewed (approved):
RSLGTM, this is absolutely incredible!

JakobJingleheimer commented on benchmark/error/node-error.js:
```diff
@@ -3,19 +3,24 @@
     },
   } = require('internal/errors');
+
+  const clazz = type === 'node' ? ERR_INVALID_STATE.TypeError : TypeError;
```
nit: constructors (and aliases to them) are usually PascalCase

```suggestion
  const Clazz = type === 'node' ? ERR_INVALID_STATE.TypeError : TypeError;
```

BridgeAR replied:
Done

JakobJingleheimer commented on benchmark/error/node-error.js:
```diff
@@ -3,19 +3,24 @@
-  for (let i = 0; i < n; ++i)
-    new ERR_INVALID_STATE.TypeError('test');
+  for (let i = 0; i < n; i++) {
+    new clazz('test');
```
```suggestion
    new Clazz('test');
```

JakobJingleheimer commented on lib/internal/streams/readable.js:
```diff
@@ -370,10 +370,10 @@ Readable.prototype.setEncoding = function(enc) {
 };
 
 // Don't raise the hwm > 1GB.
-const MAX_HWM = 0x40000000;
+const MAX_HWM = 2 ** 30;
```
nit: why not pre-compute this and annotate with the derivation?

```js
// Don't raise the hwm > 1GB (2^30).
const MAX_HWM = 1_073_741_824;
```

BridgeAR replied:
Done

JakobJingleheimer reviewed (approved):
🙌

general nit: there are several large numbers in messages for humans, but they're inconsistently delimited. for example `'The value of "port" is out of range. It must be >= 0 && <= 65535. Received 99_999'`. Why not delimit all human-friendly numbers?

mcollina commented:
@BridgeAR this now conflicts.

BridgeAR commented:
I addressed all comments as far as I can tell. I also added two more commits. The latter fixes two minor bugs.

> general nit: there are several large numbers in messages for humans, but they're inconsistently delimited.

The reason is that it's the implementers decision how to format that part. I would rather change that in another PR as this is already pretty big.

aduh95 commented on lib/internal/errors.js:
```diff
@@ -146,7 +126,27 @@ const maybeOverridePrepareStackTrace = (globalThis, error, trace) => {
+  const { toString } = error;
+  if (typeof toString === 'function' &&
+      toString !== MainContextObjectToString) {
+    errorString = FunctionPrototypeCall(toString, error);
+    if (errorString === '[Object object]' || typeof errorString !== 'string') {
+      errorString = ErrorPrototypeToString(error);
+    }
```
I think the `'[Object object]'` check is not necessary since we checked before that the function is not `%Object.prototype.toString%`. I'm suggesting replacing the `typeof` check with a template string to force it to be a string, and if any error arises, fallback to the primordials. wdyt?
```suggestion
    try {
      errorString = `${FunctionPrototypeCall(toString, error)}`;
    } catch {    
      errorString = ErrorPrototypeToString(error);
    }
```

aduh95 replied:
No strong feelings from me, it's hard to find the right balance between helpfulness and "magic" behavior, IMO if the user changes their Error objects to the `toString` method returns e.g. a number, that's on them, and maybe they expect to see a number in place of a stack trace? Or maybe they've made a mistake in their code, and won't understand why their `toString` method is called but its result is not used?

BridgeAR replied:
I decided to stick to the current behavior in this PR. I do plan on working further on the errors as there is quite a bit room for further improvements later on.

aduh95 commented on lib/internal/errors.js:
```diff
@@ -146,7 +126,27 @@ const maybeOverridePrepareStackTrace = (globalThis, error, trace) => {
+  if (typeof toString === 'function' &&
+      toString !== MainContextObjectToString) {
+    errorString = FunctionPrototypeCall(toString, error);
+    if (errorString === '[Object object]' || typeof errorString !== 'string') {
```
`Object.prototype.toString.call(new Error)` returns a different string btw:

```suggestion
    if (errorString === '[object Error]' || typeof errorString !== 'string') {
```

BridgeAR replied:
I changed the check to a more generic one.

aduh95 commented on lib/internal/errors.js:
```diff
@@ -146,7 +124,26 @@ const maybeOverridePrepareStackTrace = (globalThis, error, trace) => {
+  const { toString } = error;
+  if (typeof toString === 'function') {
+    errorString = FunctionPrototypeCall(toString, error);
+    if (typeof errorString !== 'string' || errorString.startsWith('[object ')) {
```
```suggestion
    if (typeof errorString !== 'string' || StringPrototypeStartsWith(errorString, '[object ')) {
```

ljharb commented on lib/repl.js:
```diff
@@ -646,27 +643,30 @@ function REPLServer(prompt,
-          const idx = ArrayPrototypeFindIndex(
-            ArrayPrototypeReverse(stackFrames),
-            (frame) => frame.getFunctionName() === null,
+          const idx = stackFrames.findLastIndex(
```
```suggestion
          const idx = ArrayPrototypeFindLastIndex(
            stackFrames,
```

BridgeAR replied:
This method does not exist as primordial right now.

ljharb replied:
cc @aduh95 looks like this one needs to be added then

aduh95 replied:
That's because that method can be disabled using a V8 flag:

https://github.com/nodejs/node/blob/71b9f792237b39f283eff4cd7ed35212e8669a45/deps/v8/src/flags/flag-definitions.h#L251

ljharb commented on lib/internal/errors.js:
```diff
@@ -828,16 +765,20 @@ function setArrowMessage(err, arrowMessage) {
+        }
+      }
+    } else {
+      for (const frame of stackFrames) {
```
for-of is vulnerable to Array.prototype Symbol.iterator modification, this should be a regular for loop, or ArrayPrototypeForEach

ljharb replied:
Why would you prefer this over the much faster and more robust `for (let i = 0; i < stackFrames.length; i++) { const frame = stackFrames[i];`, in a performance improvement PR?

aduh95 replied:
Last time I checked, classic `for` loops are still the most efficient way of iterating over an array, I agree with Jordan, it'd be nice to use those here.

ljharb replied:
@JakobJingleheimer it's also an issue of robustness. I don't care about while vs for - personal preference can rule there - but for-of is simply not safe like the others.

ljharb commented on lib/internal/errors.js:
```diff
@@ -828,16 +765,20 @@ function setArrowMessage(err, arrowMessage) {
-        (frm) => !StringPrototypeStartsWith(frm.getFileName() || '',
-                                            'node:internal'),
-      );
+      for (const frame of stackFrames) {
```
for-of is vulnerable to Array.prototype Symbol.iterator modification, this should be a regular for loop, or ArrayPrototypeForEach

BridgeAR replied:
We do not enforce such usage and I would like to keep them as is.

ljharb replied:
see https://github.com/nodejs/node/pull/46648#discussion_r1122557460

ljharb commented on lib/internal/errors.js:
```diff
@@ -83,55 +74,42 @@ const kTypes = [
-    if (trace.length > userStackTraceLimit)
-      ArrayPrototypeSplice(trace, userStackTraceLimit);
+    if (i === 0) {
+      trace.shift();
```
```suggestion
      ArrayPrototypeShift(trace);
```

ljharb reviewed (approved):
Looks great once the primordial issues are resolved!

mcollina reviewed (approved):
lgtm

benjamingr commented on lib/internal/assert/assertion_error.js:
```diff
@@ -449,10 +451,6 @@ class AssertionError extends Error {
     }
     ErrorCaptureStackTrace(this, stackStartFn || stackStartFunction);
-    // Create error message including the error code in the name.
-    this.stack; // eslint-disable-line no-unused-expressions
```
This was here intentionally, why was it here and why is removing it ok now?

benjamingr reviewed (commented):
I'm traveling and won't have time to review fully but please let's all be super careful landing these sort of changes and this one (changing stack creation and errors) in particular. They have a very high breakage potential.

Other than that great work Ruben!

BridgeAR commented on lib/internal/crypto/hkdf.js:
```diff
@@ -57,7 +57,7 @@ const validateParameters = hideStackFrames((hash, key, salt, info, length) => {
 
   if (info.byteLength > 1024) {
-    throw ERR_OUT_OF_RANGE(
+    throw new ERR_OUT_OF_RANGE(
```
Classes are faster than the former error creation using a function that returns an error. This is as such a mandatory change.

jasnell commented on lib/internal/errors.js:
```diff
@@ -196,21 +187,24 @@ function lazyBuffer() {
+    try {
+      Error.stackTraceLimit = limit;
+    } catch {
+      stackTraceLimitWritable = -1;
```
Nit: A comment here on why this would throw would be helpful for future generations ;-)

BridgeAR replied:
I suggest to address that in a follow-up PR? :)

```
      // Setting the property in strict mode would cause this to throw an error
      // while using the --frozen-intrinsics flag due to the writable attribute
      // being set to false.
```

BridgeAR replied:
I actually wonder, if we want to keep that property writable even when using frozen intrinsics.
@aduh95 @bmeck you might have an opinion about that?

aduh95 replied:
To fill the contract of `--forzen-intrinsics` (opt-out of all possible prototype mutation), we need the property to be non-configurable, it's true that it could still be writable without introducing too much risk I believe.

But even if we did that, we would probably still want the try catch in case some user code has changed this to non-writable for some reason, unless it's introducing too much maintenance burden. wdyt?

Uzlopak commented:
I would actually say, that we should split this PR in multiple PRs so that it can be easier reviewed.
